content/blog/2022-04-11-CML-runners-saving-models-1.md

```diff
@@ -0,0 +1,302 @@
+  always uses the latest input data.
+picture: 2022-04-11/saving-models-cover.jpeg
+# pictureComment: Some _Comment_ (supports _basic_ [Markdown](link))
+author: rob_dewit
+# commentsUrl: https://discuss.dvc.org/t/february-22-community-gems/1078
+tags:
+  - CML
+  - Git
+  - Pipelines
+  - Self-hosted runners
+  - CI/CD
+  - Cloud training
+  - AWS
+---
+
+When you first develop a machine learning model, you will probably do so on your
+local machine. You can easily change algorithms, parameters, and input data
+right in your text editor, notebook, or terminal. Imagine you have a
+long-running model for which you want to detect possible
+[drift](https://en.wikipedia.org/wiki/Concept_drift), however. In that case it
+would be beneficial to automatically retrain your model on a regular basis.
+
+In this guide, we will show how you can use
+[CML (Continuous Machine Learning)](https://cml.dev/) to do just that. CML is an
+open-source library for implementing continuous integration and delivery (CI/CD)
+in machine learning projects. This way we can define a pipeline to train a model
+and keep track of various versions. Although we could do so directly in our
+CI/CD pipeline (e.g. GitHub Workflows), the runners used for this generally
+don’t have a lot of processing power. Therefore it makes more sense to provision
+a dedicated runner that is tailored to our computing needs.
+
+At the end of this guide we will have set up a CML workflow that does the
+following on a daily basis:
+
+1. Provision an Amazon Web Services (AWS) EC2 instance
+1. Train the model
+1. Save the model and its metrics to a GitHub repository
+1. Terminate the AWS EC2 instance
+
+In a follow-up post we will expand upon this by using [DVC](https://dvc.org/) to
+designate a remote storage for our resulting models. But let's focus on CML
+first!
+
+All files needed for this guide can be found in
+[this repository](https://github.com/iterative/example_model_export_cml).
+
+<admon type="info">
+
+This guide can be followed on its own, but also as an extension to this
+[example in the docs](https://cml.dev/doc/self-hosted-runners).
+
+</admon>
+
+<admon type="tip">
+
+We wil be using GitHub for our CI/CD and AWS for our computing resources. With
+slight modifications, however, you can use Gitlab and Google Cloud or Microsoft
+Azure respectively.
+
+</admon>
+
+# Prerequisites
+
+Before we begin, make sure you have the following things set up:
+
+1. You have
+   [created an AWS account](https://aws.amazon.com/premiumsupport/knowledge-center/create-and-activate-aws-account/)
+   (free tier suffices)
+2. You have
+   [created a `PERSONAL_ACCESS_TOKEN` on GitHub](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
+3. You have
+   [created an `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` on AWS](https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html)
```

Comment:
```suggestion
3. You have
   [created an `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` on AWS](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-creds)
```

probably best? By comparison the following seem slightly worse:

- https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html
- https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html
- https://cml.dev/doc/self-hosted-runners?tab=AWS#cloud-compute-resource-credentials

WDYT? /CC @iterative/cml